### PR TITLE
Fix privileges handling

### LINKED ISSUE #243: Grant fails if symbol privilege contains an underscore
Hello

the manual implies that we can express grant privileges with symbols : 
```ruby
# Grant SELECT, UPDATE, and INSERT privileges to all tables in foo db from all hosts
mariadb_user 'foo_user' do
  password 'super_secret'
  database_name 'foo'
  host '%'
  privileges [:select,:update,:insert]
  action :grant
end
```

the full list can be found [here](https://github.com/sous-chefs/mariadb/blob/726d8cb0cfbb6d3417942a52560ffef09c0a0e6f/resources/user.rb#L165)

However MariaDB doesn't allow granting privileges containing an undersocre : 
```sql
MariaDB [(none)]> GRANT select ON *.* TO 'root'@'localhost';
Query OK, 0 rows affected (0.00 sec)

MariaDB [(none)]> GRANT show_view ON *.* TO 'root'@'localhost';
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'ON *.* TO 'root'@'localhost'' at line 1
```

So using : 
```ruby
  privileges [:show_view]
```
makes the cookbook fail.

To grant that permission, one has to write : 
```ruby
  privileges ["SHOW VIEW"]
```

I suggest the manual be updated to explicitely use a string representing the usual uppercase form.

### PR DESCRIPTION
**Pull request by dud225**
## Description

This patch fixes the handling of privileges by enabling the user to specify privileges containing spaces.

### Issues Resolved

Fixes #243

### CODE REVIEW

File: resources/user.rb [hunk 1]
```diff
@@ -220,52 +220,51 @@ def revokify_key(key)
...
+  else
+    # These should all be 'Y'
     parsed_result = parse_mysql_batch_result(test_sql_results)
     Chef::Log.debug(parsed_result)
     parsed_result.each do |r|
       desired_privs.each do |p|
         key = p.to_s.capitalize.tr(' ', '_').gsub('Replication_', 'Repl_').gsub('Create_temporary_tables', 'Create_tmp_table').gsub('Show_databases', 'Show_db')
         key = "#{key}_priv"
-        incorrect_privs = true if r[key] != 'Y'
+        privs_to_grant << get_clean_grant_right_name(p) if r[key] != 'Y'
```
Comment (sinfomicien) [inline resources/user.rb:240]: Unfortunately, it's not working with grant. You cannot add right one by one, like you can do with revoke (you can revoke only 1 right on 4). You must set all rights you want within one GRANT. Not that easy to understand... But it's like that.

So if one right is missing, you have to add all the rights you want in the GRANT request. That's why we used here a boolean.

File: resources/user.rb [hunk 2]
```diff
@@ -220,52 +220,51 @@ def revokify_key(key)
...
-      repair_sql << " TO '#{new_resource.username}'@'#{new_resource.host}' IDENTIFIED BY"
-      repair_sql << if new_resource.password.is_a?(HashedPassword)
-                      " PASSWORD '#{new_resource.password}'"
-                    else
-                      " '#{new_resource.password}'"
-                    end
-      repair_sql << ' REQUIRE SSL' if new_resource.require_ssl
-      repair_sql << ' REQUIRE X509' if new_resource.require_x509
-      repair_sql << ' WITH GRANT OPTION' if new_resource.grant_option
+      grant_statement = "GRANT #{privs_to_grant.join(',')}"
```
Comment (sinfomicien) [inline resources/user.rb:251]: Same comment as before...

**Review by sinfomicien — CHANGES_REQUESTED** [review]
One small misunderstanding to correct because of how GRANT request works. Except that point, it's what i had in mind. Good job :-)

File: resources/user.rb (hunk 1 above)
Comment (dud225) [reply resources/user.rb:240]: You mean that 2 subsequent queries, for ex. :
"GRANT SELECT"
followed by
"GRANT UPDATE"
won't work as intended ? 
What will be the final state ? Only the last one will be enforced ?

Comment (dud225) [reply resources/user.rb:240]: I not sure to properly get your point, this code only executes one GRANT request.

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by xorima** [comment]
As there seems to be some confusion on this is there any chance @dud225 you could get an integration test for this one so we can confirm the current behaviour works and your desired behaviour works (and we do not break it in future)

I see you did the spec tests as well but I think a functional test could be really useful here

Also apologies for the delay coming back to you,

**Comment by xorima** [comment]
Hi,

I am closing this due to a lack of response, however please do re-open this PR with tests for the behaviour that is not working/ this pr resolves and we can look to get this merged in

Thanks

Sous Chefs

**Pull request closed without merging**